Mage.Sets/src/mage/cards/m/Mordenkainen.java

```diff
@@ -0,0 +1,117 @@
+
+        // +2: Draw two cards, then put a card from your hand on the bottom of your library.
+        this.addAbility(new LoyaltyAbility(new MordenkainenDrawEffect(), 2));
+
+        // −2: Create a blue Dog Illusion creature token with "This creature's power and toughness are each equal to twice the number of cards in your hand."
+        this.addAbility(new LoyaltyAbility(new CreateTokenEffect(new DogIllusionToken()), -2));
+
+        // −10: Exchange your hand and library, then shuffle. You get an emblem with "You have no maximum hand size."
+        Ability ability = new LoyaltyAbility(new MordenkainenExchangeEffect(), -10);
+        ability.addEffect(new GetEmblemEffect(new MordenkainenEmblem()));
+        this.addAbility(ability);
+    }
+
+    private Mordenkainen(final Mordenkainen card) {
+        super(card);
+    }
+
+    @Override
+    public Mordenkainen copy() {
+        return new Mordenkainen(this);
+    }
+}
+
+class MordenkainenDrawEffect extends OneShotEffect {
+
+    public MordenkainenDrawEffect() {
+        super(Outcome.DrawCard);
+        this.staticText = "Draw two cards, then put a card from your hand on the bottom of your library";
+    }
+
+    private MordenkainenDrawEffect(final MordenkainenDrawEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public MordenkainenDrawEffect copy() {
+        return new MordenkainenDrawEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller == null) {
+            return false;
+        }
+        controller.drawCards(2, source, game);
+        TargetCardInHand target = new TargetCardInHand();
+        if (controller.chooseTarget(Outcome.Discard, target, source, game)) {
+            Card card = game.getCard(target.getFirstTarget());
+            if (card != null) {
+                controller.moveCardToLibraryWithInfo(card, source, game, Zone.HAND, false, false);
+            }
+        }
+        return true;
+    }
+}
+
+class MordenkainenExchangeEffect extends OneShotEffect {
+
+    public MordenkainenExchangeEffect() {
+        super(Outcome.Benefit);
+        this.staticText = "Exchange your hand and library, then shuffle";
+    }
+
+    private MordenkainenExchangeEffect(final MordenkainenExchangeEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public MordenkainenExchangeEffect copy() {
+        return new MordenkainenExchangeEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller == null) {
+            return false;
+        }
+        Cards hand = new CardsImpl(controller.getHand());
+        Cards library = new CardsImpl(controller.getLibrary().getCardList());
+        controller.moveCards(hand, Zone.LIBRARY, source, game);
```

Comment:
I'm pretty sure if you're putting cards into a library you need to use the `putCardsOnTopOfLibrary` method, but it's not something I'm super sure of

Reply:
I changed this.  They both appear to do the same thing in game.  I guess the card ordering could be different but the library is getting shuffled immediately after anyway.